### PR TITLE
Displays an error to the user if connection is lost or fetch call fails

### PR DESCRIPTION
**Pull request by HelloUniverseNo31415**
# Checklist:

Related work items: #200 

Preview url: https://hsk-gurveer.firebaseapp.com/

- [x] Tests have been added for new functionality
- [x] Work has been reviewed by other team members
- [ ] Work has been reviewed by @craigayre or @eclennett

### CODE REVIEW

File: src/components/App/App.js [hunk 1]
```diff
@@ -48,15 +57,38 @@ class App extends Component {
...
     clearInterval(this.interval);
     document.body.removeEventListener('touchstart', this.resetTimeoutTimer);
+    window.removeEventListener('online', this.handleOnline);
+    window.removeEventListener('offline', this.handleOffline);
   }
 
   render() {
     return (
-      <div>
+      <div key={this.state.refresh}>
```
Comment (isaac-lapworth) [inline src/components/App/App.js]: could use isOnline as the key and get rid of refresh from state?

File: src/components/ItemRecognition/ItemRecognition.js [hunk 2]
```diff
@@ -14,10 +14,24 @@ class ItemRecognition extends Component {
...
+    }
+  }
+
   componentDidMount() {
     this.props.setPrediction(null, null);
-    this.model.load();
-    this.controllerDataset = new ControllerDataset();
+
+    if (navigator.onLine) {
+      this.model = new Model();
```
Comment (isaac-lapworth) [inline src/components/ItemRecognition/ItemRecognition.js]: If a new Model is created here can the duplicate of this statement be removed from the constructor?

File: src/utils/retry.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const retry = async (
+  asyncRequest,
+  onInitialError,
+  totalAttempts = 10,
+  sleepTime = 2000
+) => {
+  let lastError;
```
Comment (isaac-lapworth) [inline src/utils/retry.js]: found this a little misleading, can you rename it to something like `latestError` or `previousError`?

File: src/utils/retry.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  let attempts = 0;
+
+  while (totalAttempts > attempts) {
```
Comment (isaac-lapworth) [inline src/utils/retry.js]: conventionally: `while(attempts < totalAttempts) {`

File: src/utils/retry.js [hunk 5]
```diff
@@ -4,10 +4,10 @@ const retry = async (
   totalAttempts = 10,
   sleepTime = 2000
 ) => {
-  let lastError;
+  let latestError;
   let attempts = 0;
 
-  while (totalAttempts > attempts) {
+  while (attemptstotal <Attempts) {
```
Comment (isaac-lapworth) [inline src/utils/retry.js]: typo here?

File: src/components/StoreList/actions.js [hunk 6]
```diff
@@ -17,6 +19,9 @@ export function setLoadStoreListError(loadStoreListError) {
 }
 
 export const loadStoreList = () => dispatch =>
-  fetchItems()
-    .then(storeList => dispatch(setStoreList(storeList)))
+  retry(() => fetchItems(), () => dispatch(setLoadStoreListError(true)))
```
Comment (eclennett) [inline src/components/StoreList/actions.js]: what are you passing in arrow functions...

File: src/components/App/App.js [hunk 7]
```diff
@@ -50,15 +57,37 @@ class App extends Component {
...
+      isOnline: true
+    });
+    // this isnt the nicest way of doing it but Im uysing this to reload the route, I only want to do this on the online event though, hence im not using the isOnline as a key (to router)
+  };
+
+  handleOffline = () => {
+    this.setState({isOnline: false});
+  };
+
+  connectionError() {
```
[the remark this answers is not in the document]
Comment (HelloUniverseNo31415) [reply src/components/App/App.js]: It will be true if anyone one of error  is true or isOnline  is false tight?

File: src/utils/retry.test.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import retry from './retry';
+
+test('Correctly returns what the function it is called with returns', () => {
+  retry(() => 'foo').then(result => expect(result).toEqual('foo'));
+});
+
+test('Retries if initial is an error', () => {
+  let previousCalls = false;
+  const testFunction = () => {
```
Comment (eclennett) [inline src/utils/retry.test.js]: test function is duplicated so can move it out of the scope

**Pull request merged**